### PR TITLE
SOME/IP rewrite to facilitate binding + bind SD

### LINKED ISSUE #2343: Scapy doesn't recognize SOME/IP-SD layer
#### Environment

-   Scapy version: v2.4.3.dev31
-   Python version: 3.6.4
-   Operating System: Windows 10 64bit

#### How to reproduce
```
from scapy.all import *
load_contrib("automotive.someip_sd")

ea = SDEntry_Service()
ea.tpe = 0
ea.srv_id = 0x1234
ea.inst_id = 0x5678
ea.ttl = 3

sd = SD()
sd.set_entryArray(ea)
some = sd.get_someip(True)

print("original packet:")
some.show()

some_raw = SOMEIP(some.build())

print("rebuilt packet:")
some_raw.show()
```

#### Actual result

```
original packet:
###[ SOME/IP ]### 
  \msg_id    \
   |###[ MessageId ]### 
   |  srv_id    = 65535
   |  sub_id    = EVENT_ID
   |  event_id  = 256
  len       = None
  \req_id    \
   |###[ RequestId ]### 
   |  client_id = 0
   |  session_id= 0
  proto_ver = 1
  iface_ver = 1
  msg_type  = NOTIFICATION
  retcode   = E_OK
###[ SD ]### 
     flags     = 0
     res       = 0x0
     len_entry_array= None
     \entry_array\
      |###[ Service Entry ]### 
      |  type      = 0
      |  index_1   = 0
      |  index_2   = 0
      |  n_opt_1   = 0
      |  n_opt_2   = 0
      |  srv_id    = 4660
      |  inst_id   = 22136
      |  major_ver = 0
      |  ttl       = 0x3
      |  minor_ver = 0
     len_option_array= None
     \option_array\

rebuilt packet:
###[ SOME/IP ]### 
  \msg_id    \
   |###[ MessageId ]### 
   |  srv_id    = 65535
   |  sub_id    = EVENT_ID
   |  event_id  = 256
  len       = 36
  \req_id    \
   |###[ RequestId ]### 
   |  client_id = 0
   |  session_id= 0
  proto_ver = 1
  iface_ver = 1
  msg_type  = NOTIFICATION
  retcode   = E_OK
###[ Raw ]### 
     load      = '\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x124Vx\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00'

```

#### Expected result

Scapy should recognize the SD layer, but instead there is only a Raw layer given. I understand, that there is no simple bind, because of the fieldclasses, but there should be another possibility to bind the SOMEIP and the SD layer. 

#### Related resources

Autostar writes in the official protocol docs:

> [PRS_SOMEIPSD_00151]dService Discovery messages shall use the Service-ID (16Bits) of 0xFFFF

> [PRS_SOMEIPSD_00152]dService Discovery messages shall use the Method-ID (16Bits) of 0x8100

> [PRS_SOMEIPSD_00154]dService Discovery messages shall have a Client-ID (16Bits) set to  0x0000, since there exists only a single SOME/IP-SD instance.

This could be used to recognize a SD Layer.

### PR DESCRIPTION
**Pull request by StevenVanAcker**
This PR does away with some unnecessary complexity through PacketFields which breaks bind_layers() functionality. It also combines SOME/IP and SOME/IP+SD into a single contrib module. In essence, this PR picks up where (abandoned?) PR #2347 left off, fixing Issue #2343 as well. Documentation and tests have been updated.

fixes #2343

### CODE REVIEW

**Comment by polybassa** [comment]
Thanks for your PR.

 Since you are working on SOME/IP at the moment, could you please take a look at #2348. Maybe that's easy fixable as well. 
Again here, the failed test is not related to you changes. Will hopefully be fixed with #2428 

Besides that, could you please add some minimal tests for your chances.

**Comment by StevenVanAcker** [comment]
I will give it a go :)

**Comment by StevenVanAcker** [comment]
I believe I've tracked down the problem with #2348, but it involves fixing SD entries and is not really related to this PR. I don't have a problem updating this PR with extra fixes and (if possible?) change the description of this PR. However, that may not be very clean. What is your preference: fix it in this PR or start another one?

**Comment by polybassa** [comment]
If both fixes don't depend on each other, I would prefer a new PR. Thanks!

**Comment by StevenVanAcker** [comment]
allright, will submit a new PR when it's ready :)

**Comment by polybassa** [comment]
@baarse Could you please have a look at this PR.
This PR looks good from my side.

**Comment by baarse** [comment]
Thanks for your PR, looks good from my perspective.

**Comment by StevenVanAcker** [comment]
a follow-up PR with a fix for #2348 is ready to go

File: scapy/contrib/automotive/someip.py [hunk 1]
```diff
@@ -225,3 +206,326 @@ def _bind_someip_layers():
...
+    fields_desc = [
+        _SDEntry,
+        BitField("res", 0, 12),
+        BitField("cnt", 0, 4),
+        ShortField("eventgroup_id", 0)
+    ]
+
+
+# SD Option
+#  - Configuration
```
Comment (guedou) [inline scapy/contrib/automotive/someip.py]: There are two white-spaces here and below, do you want to change that?

Comment (StevenVanAcker) [reply scapy/contrib/automotive/someip.py]: sure :)

**Comment by gpotter2** [comment]
Please remember to squash PRs when merging :-P

**Comment by daytona675** [comment]
Hi all, i've follow the issue and the PR but with the latest version of scapy and python 3.11, I have the issue.
I'm on linux.

here below a sample of scapy trace (with scapy cli)

ea = SDEntry_Service()
...: ea.type = 0
...: ea.srv_id = 0x1234
...: ea.inst_id = 0x5678
...: ea.ttl = 3
>>> someraw=SOMEIP(sd.build())
>>> someraw.show2()
###[ SOME/IP ]### 
  srv_id    = 0x0
  sub_id    = METHOD_ID
  method_id = 0x0
  len       = 16
  client_id = 0x0
  session_id= 0x0
  proto_ver = 0x12
  iface_ver = 0x34
  msg_type  = 86
  retcode   = 120
###[ Raw ]### 
     load      = '\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00'

>>> ea = SDEntry_Service()
...: ea.type = 0
...: ea.srv_id = 0x1234
...: ea.inst_id = 0x5678
...: ea.ttl = 3
>>> ea.method_id=0x8100
>>> ea.msg_type=0x0
>>> ea.proto_ver=0x0
>>> ea.iface_ver=0x1
>>> someraw=SOMEIP(sd.build())
>>> someraw.show2()
###[ SOME/IP ]### 
  srv_id    = 0x0
  sub_id    = METHOD_ID
  method_id = 0x0
  len       = 16
  client_id = 0x0
  session_id= 0x0
  proto_ver = 0x12
  iface_ver = 0x34
  msg_type  = 86
  retcode   = 120
###[ Raw ]### 
     load      = '\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00'

**Pull request merged**